### DISCUSSION TITLE
Hardcore Suspend. Single Restorable save that ends emulation on save.

### DISCUSSION THREAD

**Discussion by kdecks** [RetroAchievements/RAIntegration#757] [Ideas]
As a hardcore player I would like to be able to ``suspend`` my current hardcore session, (which would save and end the game), turn off my emulator and then later restore the session and continue playing in hardcore.

This sounds like a difficult problem. I have no idea how realistic of a request this is, and the implementation would likely vary from emulator to emulator. But, the fact that hit counts can be saved in softcore saves makes this feel more possible than ever before.

**Comment by Jamiras** [comment]
The problem isn't any more difficult than being able to create a savestate from the toolkit. The problem is that this isn't exactly "hardcore".
 
http://docs.retroachievements.org/FAQ/#what-is-hardcore-mode

> You would have to complete the game and get the achievements first time, just like it were on the original console.

**Comment by Jamiras** [comment]
You also have to consider abuse. What's to prevent someone from "suspending" their session just before something difficult, and externally backing up the suspended state so they can continuously restore it and attempt the difficult thing again? It's just a slower, more complex save state.

**Reply by SpaceRaton** [reply]
Reviving an old thread here, but I'm interested in this concept and I have an armchair-developer-y idea about the aforementioned local backup abuse. I can't conceive of any client-side solutions, but if we're not allergic to tracking one more thing in the database pre-v2, there may be a couple options. Aside from the extreme and probably unfeasible solution of literally storing the suspend-states on the server, I wonder if we could instead track _hashes_ of suspend-state files. I'm imagining a process like this:

When a user creates a suspend-state:
-Hash the local suspend-state file
-Send the hash to the server, where it's written to a table for the user's account

When a user attempts to load a suspend-state:
-Validate the suspend-state's hash with the server
-If it's valid, load the suspend-state, and tell the server to delete the hash (or flag it as "used" if we want to keep a history)
-If it's not valid, refuse to load it and display an error

Thoughts on this approach?

**Reply by wansti** [reply]
I just had the exact same idea.

Storing hashes of save states would definitely be feasible. Once a save state has been used, the corresponding hash is deleted from the server so it cannot be used again. You could still create save states for use in softcore more if you designate a save slot for use with hardcore - saving to that slot will reset the game and upload the hash.

I can think of two situations where this would be extremely useful without breaking the spirit of "playing the game as it was intended". All it prevents is having to leave your emulator running overnight (as we all had to with our 3rd and 4th gen consoles).

* Handheld gaming, in situations where you may get interrupted (similar to e.g. Virtual Console on 3DS).
* Cheevos that require you to play for a substantial amount of time before you reach the desired spot (e.g. the speedrunning cheevos in Super Mario Land where you have to either play through the entire game twice to unlock level select before attempting them, or restart the entire game each time you fail).
